File: examples/python/geo_targeting.py
"""
Geo-targeted request via Thordata proxy network.

This example shows how to send a request that appears to come from a specific
country/region (if your Thordata plan supports geo-targeting).

We call https://ipinfo.thordata.com via the Thordata proxy and instruct the
gateway to use a specific country, so that the reported IP/location reflect
that geo-targeting.

Usage (from repo root):

    # Default: country=us
    python examples/python/geo_targeting.py

    # Custom country:
    python examples/python/geo_targeting.py --country de
"""

from __future__ import annotations

import argparse
import os
from typing import Any, Dict

from dotenv import load_dotenv
from thordata import ThordataClient

load_dotenv()

SCRAPER_TOKEN = os.getenv("THORDATA_SCRAPER_TOKEN")

if not SCRAPER_TOKEN:
    raise RuntimeError("Please set THORDATA_SCRAPER_TOKEN in your .env file.")

client = ThordataClient(
    scraper_token=SCRAPER_TOKEN,
    public_token="",
    public_key="",
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Geo-targeted IP info via Thordata proxy.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument(
        "--country",
        "-c",
        type=str,
        default="us",
        help="Target country code (e.g. 'us', 'de', 'fr').",
    )
    return parser.parse_args()


def pretty_print_ipinfo(data: Dict[str, Any]) -> None:
    """
    Pretty-print a typical ipinfo-style JSON response.

    We don't rely on a strict schema here; we just try common keys.
    """
    ip = data.get("ip")
    country = data.get("country") or data.get("country_name")
    region = data.get("region") or data.get("state")
    city = data.get("city")

    print("IP info JSON:", data)
    print("\nParsed fields:")
    print(f"  IP      : {ip}")
    print(f"  Country : {country}")
    print(f"  Region  : {region}")
    print(f"  City    : {city}")


def main() -> None:
    args = parse_args()

    url = "https://ipinfo.thordata.com"
    country = args.country.lower()

    print(f"Requesting {url} via Thordata proxy (country={country})")

    # Geo-targeting is implemented on the Thordata gateway side.
    # Here we demonstrate passing 'X-Thordata-Country' as a query parameter.
    params = {"X-Thordata-Country": country}

    resp = client.get(url, params=params, timeout=30)
    resp.raise_for_status()

    data = resp.json()
    pretty_print_ipinfo(data)


if __name__ == "__main__":
    main()

File: examples/go/geo_targeting.go
// Geo-Targeted Request via Thordata Proxy
//
// Usage:
//   go run geo_targeting.go
//   go run geo_targeting.go -country de

package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
)

func main() {
	country := flag.String("country", "us", "Target country code")
	flag.Parse()

	username := os.Getenv("THORDATA_USERNAME")
	password := os.Getenv("THORDATA_PASSWORD")
	host := os.Getenv("THORDATA_PROXY_HOST")
	port := os.Getenv("THORDATA_PROXY_PORT")

	if host == "" {
		host = "pr.thordata.net"
	}
	if port == "" {
		port = "9999"
	}

	if username == "" || password == "" {
		log.Fatal("❌ Error: Set THORDATA_USERNAME and THORDATA_PASSWORD")
	}

	// Build username with geo-targeting
	proxyUsername := fmt.Sprintf("td-customer-%s-country-%s", username, *country)
	proxyURL := &url.URL{
		Scheme: "http",
		User:   url.UserPassword(proxyUsername, password),
		Host:   fmt.Sprintf("%s:%s", host, port),
	}

	fmt.Printf("🌍 Geo-targeting: %s\n", *country)
	fmt.Printf("   Username: %s\n\n", proxyUsername)

	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(proxyURL),
		},
	}

	resp, err := client.Get("https://ipinfo.io/json")
	if err != nil {
		log.Fatalf("❌ Error: %v", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		log.Fatalf("❌ Error: %v", err)
	}

	var result map[string]interface{}
	json.Unmarshal(body, &result)

	fmt.Println("✅ Response:")
	fmt.Printf("   IP:      %s\n", result["ip"])
	fmt.Printf("   Country: %s\n", result["country"])
	fmt.Printf("   Region:  %s\n", result["region"])
	fmt.Printf("   City:    %s\n", result["city"])
}